src/app/parcels/add/AddParcelForm.tsx

```diff
@@ -0,0 +1,194 @@
+    const router = useRouter();
+    const [fields, setFields] = useState(initialFields);
+    const [formErrors, setFormErrors] = useState(initialFormErrors);
+    const [submitError, setSubmitError] = useState(Errors.none);
+    const [submitErrorMessage, setSubmitErrorMessage] = useState("");
+    const [submitDisabled, setSubmitDisabled] = useState(false);
+
+    const formSections =
+        fields.shippingMethod === "Collection"
+            ? withCollectionFormSections
+            : noCollectionFormSections;
+
+    const fieldSetter = setField(setFields, fields);
+    const errorSetter = setError(setFormErrors, formErrors);
+
+    const mergeDateAndTime = (date: Date, time: Date): Date => {
+        return new Date(
+            date.getFullYear(),
+            date.getMonth(),
+            date.getDate(),
+            time.getHours(),
+            time.getMinutes()
+        );
+    };
+
+    const submitForm = async (): Promise<void> => {
+        setSubmitErrorMessage("");
+        setSubmitDisabled(true);
+        let inputError;
+        if (fields.shippingMethod === "Collection") {
+            inputError = checkErrorOnSubmit(formErrors, setFormErrors);
+        } else {
+            inputError = checkErrorOnSubmit(formErrors, setFormErrors, [
+                "voucherNumber",
+                "packingDate",
+                "timeOfDay",
+                "shippingMethod",
+            ]);
+        }
+        if (inputError) {
+            setSubmitError(Errors.submit);
+            setSubmitDisabled(false);
+            return;
+        }
+        if (fields.packingDate === null || fields.timeOfDay === null) {
+            setSubmitDisabled(false);
+            return;
+        }
+
+        const packingDateTime = mergeDateAndTime(
+            new Date(fields.packingDate),
+            new Date(fields.timeOfDay)
+        );
+
+        let collectionDateTime = new Date();
+        if (fields.shippingMethod === "Collection") {
+            if (fields.collectionDate === null || fields.collectionTime === null) {
+                setSubmitDisabled(false);
+                return;
+            }
+
+            collectionDateTime = mergeDateAndTime(
+                new Date(fields.collectionDate),
+                new Date(fields.collectionTime)
+            );
+        }
+
+        const isDelivery = fields.shippingMethod === "Delivery";
+
+        const formToAdd = {
+            client_id: props.id,
+            packing_datetime: packingDateTime.toISOString(),
+            voucher_number: fields.voucherNumber,
+            collection_centre: isDelivery ? "Delivery" : fields.collectionCentre,
+            collection_datetime: isDelivery ? null : collectionDateTime.toISOString(),
+        };
+        try {
+            await insertParcel(formToAdd);
+            router.push("/clients/");
+        } catch (error: unknown) {
+            if (error instanceof Error) {
```

Comment:
I don't think so from my limited knowledge? But also idk what else could be thrown.

Reply:
_anything_ - you can just do `throw "Hello, world!"` if you wanted. It's more what happens if we throw something which isn't an error...
I think this is better encoded as `catch (error: Error)` - if TypeScript allows it!

Reply:
So I don't understand why we would ever do that? Like surely when we throw something it would be an error. I may just not know a use case for it though. Also typescript gives this error: `Catch clause variable type annotation must be 'any' or 'unknown' if specified`